### PR TITLE
Use allow-list for APM agent settings

### PR DESCRIPTION
**Pull request by mosche**
Prevent invalid and miss-configuration of APM agent using an explicit allow-list of setting keys.
Additionally this removes the redundant `APM_AGENT_DEFAULT_SETTINGS` in favor of a single definition of defaults in `APMJvmOptions`.
(Closes ES-6916)

### CODE REVIEW

File: modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java [hunk 1]
```diff
@@ -114,15 +115,130 @@ public void setAgentSetting(String key, String value) {
...
+        "ignore_exceptions",
+        "capture_body",
+        "capture_headers",
+        "global_labels",
+        "instrument_ancient_bytecode",
+        "context_propagation_only",
+        "classes_excluded_from_instrumentation",
+        "trace_methods",
+        "trace_methods_duration_threshold",
+        "central_config", // TODO TBD disabled in APMJvmOptions defaults, does this impact usage of config_file?
```
Comment (mosche) [inline modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java:156]: Not sure about this one, could it impact usage of / conflict with `config_file`?

File: modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java [hunk 2]
```diff
@@ -114,15 +115,130 @@ public void setAgentSetting(String key, String value) {
...
+        // forbid 'service_version', forced by APMJvmOptions
+        "hostname",
+        "environment",
+        "transaction_sample_rate",
+        "transaction_max_spans",
+        "long_field_max_length",
+        "sanitize_field_names",
+        "enable_instrumentations",
+        "disable_instrumentations",
+        "enable_experimental_instrumentations", // TODO TBD allow disabling this, conflicts with APMJvmOptions
```
[the remark this answers is not in the document]
Comment (mosche) [reply modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java:145]: Here's another PR to apply defaults in a single consistent location:
https://github.com/elastic/elasticsearch/pull/104149/

[the remark this answers is not in the document]
Comment (stu-elastic) [reply modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java:145]: Yeah it looks like `APM_AGENT_DEFAULT_SETTINGS` should be nuked.

File: modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java (hunk 1 above)
Comment (stu-elastic) [reply modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java:156]: Let's remove this one as it impacts `config_file`.

File: modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java [hunk 3]
```diff
@@ -114,15 +115,130 @@ public void setAgentSetting(String key, String value) {
...
+        "api_request_size",
+        "metrics_interval",
+        "disable_metrics",
+
+        // Serverless:
+        "aws_lambda_handler",
+        "data_flush_timeout",
+
+        // Stacktraces:
+        "application_packages", // TODO TBD default configured by APMJvmOptions, allow overwrites?
```
[the remark this answers is not in the document]
Comment (stu-elastic) [reply modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java:239]: We can allow it for now.

File: modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java [hunk 4]
```diff
@@ -132,7 +248,7 @@ public void setAgentSetting(String key, String value) {
             final String key = parts[parts.length - 1];
             final String defaultValue = APM_AGENT_DEFAULT_SETTINGS.getOrDefault(key, "");
             return new Setting<>(qualifiedKey, defaultValue, (value) -> {
-                if (PROHIBITED_AGENT_KEYS.contains(key)) {
+                if (qualifiedKey.equals("_na_") == false && PERMITTED_AGENT_KEYS.contains(key) == false) {
```
Comment (stu-elastic) [inline modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java:251]: I think the `_na_` had to do with `Setting.affixKeySetting`'s `delegateFactory` which is used via `Setting.prefixKeySetting` by `APM_AGENT_SETTINGS`.  Are you sure it's ok to remove this check?

Comment (mosche) [reply modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java:229]: Moving from deny to allow list, I had to explicitly allow `_na_`. As you mentioned the `delegateFactory` is always invoked once with `_na_`, which obviously shouldn't fail.

File: modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mosche) [reply modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java]: 👍 I dropped `APM_AGENT_DEFAULT_SETTINGS` and removed `enable_experimental_instrumentations` from the allow list.

File: modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java (hunk 1 above)
Comment (mosche) [reply modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java]: 👍

File: modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mosche) [reply modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java]: 👍

File: build-tools/src/main/java/org/elasticsearch/gradle/testclusters/RunTask.java [hunk 5]
```diff
@@ -213,8 +213,8 @@ public void beforeStart() {
                 // if metrics were not enabled explicitly for gradlew run we should disable them
                 else if (node.getSettingKeys().contains("telemetry.metrics.enabled") == false) { // metrics
                     node.setting("telemetry.metrics.enabled", "false");
-                } else if (node.getSettingKeys().contains("tracing.apm.agent.enabled") == false) { // tracing
-                    node.setting("tracing.apm.agent.enable", "false");
+                } else if (node.getSettingKeys().contains("tracing.apm.enabled") == false) { // tracing
+                    node.setting("tracing.apm.enable", "false");
```
Comment (pgomulka) [inline build-tools/src/main/java/org/elasticsearch/gradle/testclusters/RunTask.java:217]: turns out this is not an affix seting and we could deprecate it and use 'telemetry.tracing.enabled' right?

**Review by pgomulka — COMMENTED** [review]
I left few comments. I think they are not really related to the exact thing in this PR (allow-list of settings), but perhaps we could follow up in a separate PR

File: build-tools/src/main/java/org/elasticsearch/gradle/testclusters/RunTask.java (hunk 5 above)
Comment (mosche) [reply build-tools/src/main/java/org/elasticsearch/gradle/testclusters/RunTask.java:217]: Yes, I noticed that when talking to Stu, I'll work on that next 👍

File: modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java [hunk 6]
```diff
@@ -114,26 +94,140 @@ public void setAgentSetting(String key, String value) {
     private static final String APM_SETTING_PREFIX = "tracing.apm.";
```
[the remark this answers is not in the document]
Comment (mosche) [reply modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java:94]: Can we postpone that for a later PR when cleaning up metric names? Likely that will touch all of this anyways.

File: modules/apm/src/test/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettingsTests.java [hunk 7]
```diff
@@ -40,48 +40,29 @@ public void test_whenTracerDisabled_setsRecordingProperty() {
...
     }
 
     /**
-     * Check that when cluster settings are synchronised with the system properties, agent settings other
-     * than those with default values are set.
+     * Check that invalid or forbidden APM agent settings are rejected.
      */
-    public void test_whenTracerCreated_clusterSettingsAlsoApplied() {
-        APMAgentSettings apmAgentSettings = spy(new APMAgentSettings());
+    public void test_invalidSettingsAreRejected() {
```
[the remark this answers is not in the document]
Comment (mosche) [reply modules/apm/src/test/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettingsTests.java]: Done

**Comment by mosche** [comment]
@elasticmachine update branch

**Review by pgomulka — APPROVED** [review]
LGTM, thank you

**Pull request merged**